### PR TITLE
Documentation: Document ingester read path limiting

### PR DESCRIPTION
**Pull request by aknuds1**
#### What this PR does
Add technical documentation on resource utilization based ingester read path limiting.

#### Checklist

- [na] Tests updated.
- [x] Documentation added.
- [na] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [na] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+menuTitle: Resource utilization based ingester read path limiting
+title: Configure resource utilization based ingester read path limiting
+weight: 120
+---
+
+# Configure resource utilization based ingester read path limiting
+
+As an **experimental** feature, Grafana Mimir allows you to configure limits for respectively CPU and/or memory utilization at
+which point the ingester should reject incoming read requests, to protect the write path. The ingester's model of memory utilization
+corresponds to the Go heap size, and it tracks (a sliding window average of) the process' CPU utilization.
```
Comment (jhalterman) [inline docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md:14]: A phrasing suggestion:

```suggestion
As an **experimental** feature, Grafana Mimir allows you to configure limits for CPU and/or memory utilization which can cause the ingester to reject incoming read requests in order to protect the write path. The ingester's model of memory utilization
corresponds to the Go heap size, which it tracks, along with a sliding window average of the process' CPU utilization.
```

File: docs/sources/mimir/references/architecture/components/ingester.md [hunk 2]
```diff
@@ -95,3 +95,20 @@ For more information on shuffle sharding, refer to [Configuring shuffle sharding
...
 
 For more information about out-of-order samples ingestion, refer to [Configuring out of order samples ingestion]({{< relref "../../../configure/configure-out-of-order-samples-ingestion" >}}).
+
+## Resource utilization based ingester read path limiting
+
+Resource utilization based ingester read path limiting means the ingester can be configured to limit read requests based
+on resource (CPU/memory) utilization, to maintain a minimum of headroom for the write path to function.
+The idea is to prevent intensive queries from interrupting the write path.
+
+The process' CPU utilization and Go memory heap size are sampled every second, and a sliding window average taken of
```
Comment (jhalterman) [inline docs/sources/mimir/references/architecture/components/ingester.md:105]:
```suggestion
The process' CPU utilization and Go memory heap size are sampled every second, and a sliding window average is taken of
```

File: docs/sources/mimir/references/architecture/components/ingester.md [hunk 3]
```diff
@@ -95,3 +95,20 @@ For more information on shuffle sharding, refer to [Configuring shuffle sharding
...
+the CPU utilization. If either is greater than or equal to the corresponding configured limit, ingester read requests
+get rejected with HTTP status code 503 (Service Unavailable), until utilization levels are again below respective limits.
```
Comment (jhalterman) [inline docs/sources/mimir/references/architecture/components/ingester.md:107]:
```suggestion
get rejected with HTTP status code 503 (Service Unavailable), until utilization levels are below respective limits again.
```

File: docs/sources/mimir/references/architecture/components/ingester.md [hunk 4]
```diff
@@ -95,3 +95,20 @@ For more information on shuffle sharding, refer to [Configuring shuffle sharding
...
+
+Whenever the ingester rejects a read request due to utilization based limiting, it increments the
+`cortex_ingester_utilization_limited_read_requests_total` counter metric.
+CPU and memory utilization are also tracked, via respectively the `utilization_limiter_current_cpu_load` and
+`utilization_limiter_current_memory_usage_bytes` gauge metrics.
```
Comment (jhalterman) [inline docs/sources/mimir/references/architecture/components/ingester.md:112]:
```suggestion
CPU and memory utilization are also tracked, via the `utilization_limiter_current_cpu_load` and
`utilization_limiter_current_memory_usage_bytes` gauge metrics respectively.
```

File: docs/sources/mimir/references/architecture/components/ingester.md [hunk 5]
```diff
@@ -95,3 +95,20 @@ For more information on shuffle sharding, refer to [Configuring shuffle sharding
...
+on resource (CPU/memory) utilization, to maintain a minimum of headroom for the write path to function.
+The idea is to prevent intensive queries from interrupting the write path.
+
+The process' CPU utilization and Go memory heap size are sampled every second, and a sliding window average is taken of
+the CPU utilization. If either is greater than or equal to the corresponding configured limit, ingester read requests
+get rejected with HTTP status code 503 (Service Unavailable), until utilization levels are below respective limits again.
+
+Whenever the ingester rejects a read request due to utilization based limiting, it increments the
+`cortex_ingester_utilization_limited_read_requests_total` counter metric.
+CPU and memory utilization are also tracked, via the `utilization_limiter_current_cpu_load` and
```
Comment (pracucci) [inline docs/sources/mimir/references/architecture/components/ingester.md:111]:
```suggestion
CPU and memory utilization are also tracked, via the `cortex_ingester_utilization_limiter_current_cpu_load` and
```

File: docs/sources/mimir/references/architecture/components/ingester.md [hunk 6]
```diff
@@ -95,3 +95,20 @@ For more information on shuffle sharding, refer to [Configuring shuffle sharding
...
+`utilization_limiter_current_memory_usage_bytes` gauge metrics respectively.
```
Comment (pracucci) [inline docs/sources/mimir/references/architecture/components/ingester.md:112]:
```suggestion
`cortex_ingester_utilization_limiter_current_memory_usage_bytes` gauge metrics respectively.
```

File: docs/sources/mimir/manage/run-production-environment/production-tips/index.md [hunk 7]
```diff
@@ -41,6 +41,16 @@ On the read path, the ingester reads from the series whose chunks have already b
 
 For these reasons, run the ingesters on disks such as SSDs that have fast disk speed.
 
+### Resource utilization based ingester read path limiting
+
+The ingester supports limiting read requests based on resource (CPU/memory) utilization, in order to protect the write path.
+The ingester write path is generally considered more important than the read path in production, so it's (often) better to
+limit read requests when ingesters are under pressure than to fail writes (or even crash).
+
+While it is still an **experimental** feature, we can tentatively recommend enabling resource utilization based ingester read path limiting,
```
Comment (pracucci) [inline docs/sources/mimir/manage/run-production-environment/production-tips/index.md:50]: [non blocking] I feel that whenever we'll make it stable, we'll forget to update the doc here. I'm tempted to remove the "while it is still an experimental feature" part.

File: docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  - ../operators-guide/configure/configure-resource-utilization-based-ingester-read-path-limiting/
+description: Learn how to configure Grafana Mimir for resource utilization based ingester read path limiting.
+menuTitle: Resource utilization based ingester read path limiting
+title: Configure resource utilization based ingester read path limiting
+weight: 120
+---
+
+# Configure resource utilization based ingester read path limiting
+
+As an **experimental** feature, Grafana Mimir allows you to configure limits for CPU and/or memory utilization which can cause the ingester to reject incoming read requests in order to protect the write path. The ingester's model of memory utilization
```
Comment (pracucci) [inline docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md:12]: [non blocking] Same comment about experimental here (see the other comment).

File: docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
+---
+aliases:
+  - ../operators-guide/configure/configure-resource-utilization-based-ingester-read-path-limiting/
+description: Learn how to configure Grafana Mimir for resource utilization based ingester read path limiting.
+menuTitle: Resource utilization based ingester read path limiting
+title: Configure resource utilization based ingester read path limiting
+weight: 120
+---
+
+# Configure resource utilization based ingester read path limiting
```
Comment (pracucci) [inline docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md:10]: This doc page lacks some important information (e.g. metrics) which are documented in `docs/sources/mimir/references/architecture/components/ingester.md` instead. I think we should document the feature only in 1 place, either here or in `docs/sources/mimir/references/architecture/components/ingester.md`. Given it's not a critical feature (I mean something you absolutely need to be aware of) I would suggest to move all the doc here, and completely remove it from `docs/sources/mimir/references/architecture/components/ingester.md` (it's already listed among the list of production tips, which is good).

**Review by pracucci — APPROVED** [review]
Good from a technical correctness perspective (modulo a typo in the metric names). I left some suggestions about the doc organization.

File: docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md (hunk 9 above)
Comment (aknuds1) [reply docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md:8]: Done.

File: docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md (hunk 8 above)
Comment (aknuds1) [reply docs/sources/mimir/configure/configure-resource-utilization-based-ingester-read-path-limiting.md]: Donie, thanks.

File: docs/sources/mimir/manage/run-production-environment/production-tips/index.md (hunk 7 above)
Comment (aknuds1) [reply docs/sources/mimir/manage/run-production-environment/production-tips/index.md]: Done, thanks.

**Pull request merged**